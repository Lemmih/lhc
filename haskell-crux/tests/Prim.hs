{-# LANGUAGE MagicHash, ForeignFunctionInterface #-}
{-# LANGUAGE UnboxedTuples #-}
module LHC.Prim
    ( IO
    , Unit(Unit)
    , I8
    , I32
    , I64
    , Addr
    , bindIO
    , thenIO
    , return
    , unsafePerformIO
    , List(Nil,Cons)
    -- , sequence_
    , unpackString#
    , putStr
    , putStrLn
    , getChar
    , getLine
    , Char(C#)
    , Bool(True,False)
    , Int(I#)
    , not
    , otherwise
    , (+), (-), (*)
    -- , udiv, urem
    , sdiv, srem
    , mapM_
    , (<=), le#
    , max
    , i32toi64
    , i64toi32
    , length
    , emptyString
    ) where

-- XXX: These have kind # but it is not checked anywhere.
--      Use them incorrectly and the world will explode.
data I8
data I32
data I64
data Addr a

data Int32 = Int32 I32

data Int = I# I32

data Char = C# I32

data RealWorld#

foreign import ccall realworld# :: RealWorld#

-- data IOUnit a = IOUnit a RealWorld
--newtype IO a = IO { unIO :: RealWorld -> IOUnit a }
newtype IO a = IO (RealWorld# -> (# RealWorld#, a #))

data Unit = Unit
data List a = Nil | Cons a (List a)

unIO :: IO a -> RealWorld# -> (# RealWorld#, a #)
unIO action =
  case action of
    IO unit -> unit

bindIO :: IO a -> (a -> IO b) -> IO b
bindIO f g = IO (\s ->
  case unIO f s of
    (# s', a #) -> unIO (g a) s')

thenIO :: IO a -> IO b -> IO b
thenIO a b = bindIO a (\c -> b)

-- infixl 1 `thenIO`

return :: a -> IO a
return a = IO (\s -> (# s, a #))

unsafePerformIO :: IO a -> a
unsafePerformIO action =
  case unIO action realworld# of
    (# st, val #) -> val

foreign import ccall unsafe "putchar" c_putchar :: I32 -> IO Int32

-- sequence_ :: List (IO Unit) -> IO Unit
-- sequence_ lst =
--   case lst of
--     Nil -> return Unit
--     Cons head tail -> thenIO head (sequence_ tail)

foreign import ccall unsafe indexI8# :: Addr I8 -> I8
foreign import ccall addrAdd# :: Addr I8 -> I64 -> Addr I8
foreign import ccall "cast" i64toi32 :: I64 -> I32
foreign import ccall "cast" i8toi64 :: I8 -> I64
foreign import ccall "cast" i8toi32 :: I8 -> I32
foreign import ccall "cast" i32toi64 :: I32 -> I64

unpackString# :: Addr I8 -> [Char]
unpackString# ptr =
  case i8toi64 (indexI8# ptr) of
    0#   -> []
    char -> (:) (C# (i64toi32 char)) (unpackString# (addrAdd# ptr 1#))

foreign import ccall "getchar" c_getchar :: IO Int32

getChar :: IO Char
getChar =
  c_getchar `bindIO` \c ->
  case c of
    Int32 c# -> return (C# c#)

getLine :: IO [Char]
getLine =
  getChar `bindIO` \c ->
  case c of
    C# c# ->
      case c# of
        '\n'# -> return []
        _     -> getLine `bindIO` \cs -> return (c : cs)

putStr :: [Char] -> IO ()
putStr lst =
  case lst of
    [] -> return ()
    head : tail ->
      case head of
        C# char -> do
          c_putchar char
          putStr tail

putStrLn :: [Char] -> IO ()
putStrLn msg = putStr msg `thenIO` putStr (unpackString# "\n"#)

data Bool = False | True

not :: Bool -> Bool
not False = True
not True  = False

otherwise :: Bool
otherwise = True

foreign import ccall unsafe (+#) :: I32 -> I32 -> I32

foreign import ccall unsafe (-#) :: I32 -> I32 -> I32

foreign import ccall unsafe (*#) :: I32 -> I32 -> I32

(+) :: Int -> Int -> Int
I# a# + I# b# = I# (a# +# b#)

(-) :: Int -> Int -> Int
I# a# - I# b# = I# (a# -# b#)

(*) :: Int -> Int -> Int
I# a# * I# b# = I# (a# *# b#)

foreign import ccall unsafe udiv# :: I32 -> I32 -> I32
foreign import ccall unsafe sdiv# :: I32 -> I32 -> I32

foreign import ccall unsafe urem# :: I32 -> I32 -> I32
foreign import ccall unsafe srem# :: I32 -> I32 -> I32

-- udiv :: Int -> Int -> Int
-- udiv (I# a#) (I# b#) = I# (udiv# a# b#)

sdiv :: Int -> Int -> Int
sdiv (I# a#) (I# b#) = I# (sdiv# a# b#)

-- urem :: Int -> Int -> Int
-- urem (I# a#) (I# b#) = I# (urem# a# b#)

srem :: Int -> Int -> Int
srem (I# a#) (I# b#) = I# (srem# a# b#)

--seq :: a -> b -> b
--seq a b = b

--bit :: Int -> Int
--bit = bit
-- bit n = 1<<n

mapM_ :: (a -> IO ()) -> [a] -> IO ()
mapM_ fn lst =
  case lst of
    [] -> return ()
    (:) x xs -> fn x `thenIO` mapM_ fn xs

foreign import ccall unsafe le# :: I32 -> I32 -> I32

-- <=
(<=) :: Int -> Int -> Bool
I# a# <= I# b# =
  case le# a# b# of
    '\0'# -> False
    _     -> True

max :: Int -> Int -> Int
max a b =
  case a <= b of
    True -> b
    False -> a


length :: [a] -> Int
length lst = length_go 0 lst

length_go :: Int -> [a] -> Int
length_go n [] = n
length_go (I# n) (x:xs) = length_go (I# n + 1) xs

emptyString :: [Char]
emptyString = []

id :: a -> a
id x = x
