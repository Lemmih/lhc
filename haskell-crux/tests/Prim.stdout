node @.LHC.Prim.IO:(* →  *)
node @.LHC.Prim.I8:*
node @.LHC.Prim.I32:*
node @.LHC.Prim.I64:*
node @.LHC.Prim.Addr:(* →  *)
node @.LHC.Prim.Int32:*
node LHC.Prim.Int32:(LHC.Prim.I32 →  LHC.Prim.Int32)
node @.LHC.Prim.Int:*
node LHC.Prim.I#:(LHC.Prim.I32 →  LHC.Prim.Int)
node @.LHC.Prim.Char:*
node LHC.Prim.C#:(LHC.Prim.I32 →  LHC.Prim.Char)
node @.LHC.Prim.RealWorld#:*
node @.LHC.Prim.Unit:*
node LHC.Prim.Unit:LHC.Prim.Unit
node @.LHC.Prim.List:(* →  *)
node LHC.Prim.Nil:(∀ a. LHC.Prim.List a)
node LHC.Prim.Cons:(∀ a. a →  LHC.Prim.List a →  LHC.Prim.List a)
node @.LHC.Prim.Bool:*
node LHC.Prim.False:LHC.Prim.Bool
node LHC.Prim.True:LHC.Prim.Bool
newtype LHC.Prim.IO:(∀ a. (LHC.Prim.RealWorld# →  (# LHC.Prim.RealWorld#, a #)) →  LHC.Prim.IO a)
LHC.Prim.realworld# : LHC.Prim.RealWorld#
LHC.Prim.realworld# =
  void
LHC.Prim.unIO : ∀ sk_0_a. LHC.Prim.IO sk_0_a →  LHC.Prim.RealWorld# →  (# LHC.Prim.RealWorld#, sk_0_a #)
LHC.Prim.unIO =
  λ @.sk_0_a^1:* LHC.Prim.action^2:(LHC.Prim.IO sk_0_a) → 
  LHC.Prim.action^2
LHC.Prim.bindIO : ∀ sk_5_a sk_6_b. LHC.Prim.IO sk_5_a →  (sk_5_a →  LHC.Prim.IO sk_6_b) →  LHC.Prim.IO sk_6_b
LHC.Prim.bindIO =
  λ @.sk_5_a^3:* @.sk_6_b^4:* LHC.Prim.f^5:(LHC.Prim.IO sk_5_a) LHC.Prim.g^6:(sk_5_a →  LHC.Prim.IO sk_6_b) LHC.Prim.s^7:LHC.Prim.RealWorld# → 
  scrut^8 ← LHC.Prim.unIO
                  @.sk_5_a^3 LHC.Prim.f^5
              LHC.Prim.s^7
  (# LHC.Prim.s'^9, LHC.Prim.a^10 #) ← scrut^8:(# LHC.Prim.RealWorld#, a #)
  LHC.Prim.unIO @.sk_6_b^4
      (LHC.Prim.g^6 LHC.Prim.a^10)
    LHC.Prim.s'^9
LHC.Prim.thenIO : ∀ sk_13_a sk_14_b. LHC.Prim.IO sk_13_a →  LHC.Prim.IO sk_14_b →  LHC.Prim.IO sk_14_b
LHC.Prim.thenIO =
  λ @.sk_13_a^11:* @.sk_14_b^12:* LHC.Prim.a^13:(LHC.Prim.IO sk_13_a) LHC.Prim.b^14:(LHC.Prim.IO sk_14_b) → 
  LHC.Prim.bindIO @.sk_13_a^11
        @.sk_14_b^12 LHC.Prim.a^13
    (λ LHC.Prim.c^15:sk_13_a → 
    LHC.Prim.b^14)
LHC.Prim.return : ∀ sk_20_a. sk_20_a →  LHC.Prim.IO sk_20_a
LHC.Prim.return =
  λ @.sk_20_a^16:* LHC.Prim.a^17:sk_20_a LHC.Prim.s^18:LHC.Prim.RealWorld# → 
  (# LHC.Prim.s^18, LHC.Prim.a^17 #)
LHC.Prim.unsafePerformIO : ∀ sk_24_a. LHC.Prim.IO sk_24_a →  sk_24_a
LHC.Prim.unsafePerformIO =
  λ @.sk_24_a^19:* LHC.Prim.action^20:(LHC.Prim.IO sk_24_a) → 
  scrut^21 ← LHC.Prim.unIO
                   @.sk_24_a^19 LHC.Prim.action^20
               LHC.Prim.realworld#
  (# LHC.Prim.st^22, LHC.Prim.val^23 #) ← scrut^21:(# LHC.Prim.RealWorld#, a #)
  LHC.Prim.val^23
LHC.Prim.c_putchar : LHC.Prim.I32 →  LHC.Prim.IO LHC.Prim.Int32
LHC.Prim.c_putchar =
  λ arg^24:LHC.Prim.I32 s^25:LHC.Prim.RealWorld# → 
  primOut^26:LHC.Prim.I32, s'^27 ← external putchar arg^24
  (# s'^27, LHC.Prim.Int32
    primOut^26 #)
LHC.Prim.indexI8# : LHC.Prim.Addr LHC.Prim.I8 →  LHC.Prim.I8
LHC.Prim.indexI8# =
  λ arg^28:(LHC.Prim.Addr LHC.Prim.I8) → 
  primOut^29:LHC.Prim.I8 ← external indexI8# arg^28
  primOut^29
LHC.Prim.addrAdd# : LHC.Prim.Addr LHC.Prim.I8 →  LHC.Prim.I64 →  LHC.Prim.Addr LHC.Prim.I8
LHC.Prim.addrAdd# =
  λ arg^30:(LHC.Prim.Addr LHC.Prim.I8) arg^31:LHC.Prim.I64 → 
  primOut^32:(LHC.Prim.Addr LHC.Prim.I8) ← external addrAdd# arg^30 arg^31
  primOut^32
LHC.Prim.i64toi32 : LHC.Prim.I64 →  LHC.Prim.I32
LHC.Prim.i64toi32 =
  λ arg^33:LHC.Prim.I64 → 
  (arg^33 ::: LHC.Prim.I32)
LHC.Prim.i8toi64 : LHC.Prim.I8 →  LHC.Prim.I64
LHC.Prim.i8toi64 =
  λ arg^34:LHC.Prim.I8 → 
  (arg^34 ::: LHC.Prim.I64)
LHC.Prim.i8toi32 : LHC.Prim.I8 →  LHC.Prim.I32
LHC.Prim.i8toi32 =
  λ arg^35:LHC.Prim.I8 → 
  (arg^35 ::: LHC.Prim.I32)
LHC.Prim.i32toi64 : LHC.Prim.I32 →  LHC.Prim.I64
LHC.Prim.i32toi64 =
  λ arg^36:LHC.Prim.I32 → 
  (arg^36 ::: LHC.Prim.I64)
LHC.Prim.unpackString# : LHC.Prim.Addr LHC.Prim.I8 →  [LHC.Prim.Char]
LHC.Prim.unpackString# =
  λ LHC.Prim.ptr^37:(LHC.Prim.Addr LHC.Prim.I8) → 
  case LHC.Prim.i8toi64
           (LHC.Prim.indexI8#
             LHC.Prim.ptr^37) of
    scrut^38:LHC.Prim.I64
    0 → 
      LHC.Prim.Nil
    DEFAULT → 
      let LHC.Prim.char^39:LHC.Prim.I64 = scrut^38
      LHC.Prim.Cons (LHC.Prim.C#
            (LHC.Prim.i64toi32
              LHC.Prim.char^39))
        (LHC.Prim.unpackString#
          (LHC.Prim.addrAdd#
              LHC.Prim.ptr^37 1))
LHC.Prim.c_getchar : LHC.Prim.IO LHC.Prim.Int32
LHC.Prim.c_getchar =
  λ s^40:LHC.Prim.RealWorld# → 
  primOut^41:LHC.Prim.I32, s'^42 ← external getchar 
  (# s'^42, LHC.Prim.Int32
    primOut^41 #)
LHC.Prim.getChar : LHC.Prim.IO LHC.Prim.Char
LHC.Prim.getChar =
  LHC.Prim.bindIO
          @.LHC.Prim.Int32 @.LHC.Prim.Char
      LHC.Prim.c_getchar
    (λ LHC.Prim.c^43:LHC.Prim.Int32 → 
    scrut^44 ← LHC.Prim.c^43
    LHC.Prim.Int32 LHC.Prim.c#^45:LHC.Prim.I32 ← scrut^44:LHC.Prim.Int32
    LHC.Prim.return @.LHC.Prim.Char
      (LHC.Prim.C# LHC.Prim.c#^45))
LHC.Prim.getLine : LHC.Prim.IO [LHC.Prim.Char]
LHC.Prim.getLine =
  LHC.Prim.bindIO
          @.LHC.Prim.Char (@.LHC.Prim.List
          @.LHC.Prim.Char)
      LHC.Prim.getChar
    (λ LHC.Prim.c^46:LHC.Prim.Char → 
    scrut^47 ← LHC.Prim.c^46
    LHC.Prim.C# LHC.Prim.c#^48:LHC.Prim.I32 ← scrut^47:LHC.Prim.Char
    case LHC.Prim.c#^48 of
      scrut^49:LHC.Prim.I32
      '\n' → 
        LHC.Prim.return
            (@.LHC.Prim.List
              @.LHC.Prim.Char) LHC.Prim.Nil
      DEFAULT → 
        LHC.Prim.bindIO
                (@.LHC.Prim.List
                  @.LHC.Prim.Char)
              (@.LHC.Prim.List
                @.LHC.Prim.Char)
            LHC.Prim.getLine
          (λ LHC.Prim.cs^50:[LHC.Prim.Char] → 
          LHC.Prim.return (@.LHC.Prim.List
                @.LHC.Prim.Char) (LHC.Prim.Cons
                LHC.Prim.c^46 LHC.Prim.cs^50)))
LHC.Prim.putStr : [LHC.Prim.Char] →  LHC.Prim.IO ()
LHC.Prim.putStr =
  λ LHC.Prim.lst^51:[LHC.Prim.Char] → 
  case LHC.Prim.lst^51 of
    scrut^52:[LHC.Prim.Char]
    LHC.Prim.Nil  → 
      LHC.Prim.return
          @.LHC.Prim.Unit LHC.Prim.Unit
    LHC.Prim.Cons LHC.Prim.head^53:LHC.Prim.Char LHC.Prim.tail^54:[LHC.Prim.Char] → 
      scrut^55 ← LHC.Prim.head^53
      LHC.Prim.C# LHC.Prim.char^56:LHC.Prim.I32 ← scrut^55:LHC.Prim.Char
      LHC.Prim.thenIO @.LHC.Prim.Int32
            @.LHC.Prim.Unit
          (LHC.Prim.c_putchar
            LHC.Prim.char^56)
        (LHC.Prim.putStr
          LHC.Prim.tail^54)
LHC.Prim.putStrLn : [LHC.Prim.Char] →  LHC.Prim.IO ()
LHC.Prim.putStrLn =
  λ LHC.Prim.msg^57:[LHC.Prim.Char] → 
  LHC.Prim.thenIO @.LHC.Prim.Unit
        @.LHC.Prim.Unit (LHC.Prim.putStr
        LHC.Prim.msg^57)
    (LHC.Prim.putStr
      (LHC.Prim.unpackString# "\n"))
LHC.Prim.not : LHC.Prim.Bool →  LHC.Prim.Bool
LHC.Prim.not =
  λ arg^58:LHC.Prim.Bool → 
  case arg^58 of
    scrut^59:LHC.Prim.Bool
    LHC.Prim.False  → 
      LHC.Prim.True
    LHC.Prim.True  → 
      LHC.Prim.False
LHC.Prim.otherwise : LHC.Prim.Bool
LHC.Prim.otherwise =
  LHC.Prim.True
LHC.Prim.+# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.+# =
  λ arg^60:LHC.Prim.I32 arg^61:LHC.Prim.I32 → 
  primOut^62:LHC.Prim.I32 ← external +# arg^60 arg^61
  primOut^62
LHC.Prim.-# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.-# =
  λ arg^63:LHC.Prim.I32 arg^64:LHC.Prim.I32 → 
  primOut^65:LHC.Prim.I32 ← external -# arg^63 arg^64
  primOut^65
LHC.Prim.*# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.*# =
  λ arg^66:LHC.Prim.I32 arg^67:LHC.Prim.I32 → 
  primOut^68:LHC.Prim.I32 ← external *# arg^66 arg^67
  primOut^68
LHC.Prim.+ : LHC.Prim.Int →  LHC.Prim.Int →  LHC.Prim.Int
LHC.Prim.+ =
  λ arg^69:LHC.Prim.Int arg^70:LHC.Prim.Int → 
  scrut^71 ← arg^69
  LHC.Prim.I# LHC.Prim.a#^72:LHC.Prim.I32 ← scrut^71:LHC.Prim.Int
  scrut^73 ← arg^70
  LHC.Prim.I# LHC.Prim.b#^74:LHC.Prim.I32 ← scrut^73:LHC.Prim.Int
  LHC.Prim.I# (LHC.Prim.+#
        LHC.Prim.a#^72 LHC.Prim.b#^74)
LHC.Prim.- : LHC.Prim.Int →  LHC.Prim.Int →  LHC.Prim.Int
LHC.Prim.- =
  λ arg^75:LHC.Prim.Int arg^76:LHC.Prim.Int → 
  scrut^77 ← arg^75
  LHC.Prim.I# LHC.Prim.a#^78:LHC.Prim.I32 ← scrut^77:LHC.Prim.Int
  scrut^79 ← arg^76
  LHC.Prim.I# LHC.Prim.b#^80:LHC.Prim.I32 ← scrut^79:LHC.Prim.Int
  LHC.Prim.I# (LHC.Prim.-#
        LHC.Prim.a#^78 LHC.Prim.b#^80)
LHC.Prim.* : LHC.Prim.Int →  LHC.Prim.Int →  LHC.Prim.Int
LHC.Prim.* =
  λ arg^81:LHC.Prim.Int arg^82:LHC.Prim.Int → 
  scrut^83 ← arg^81
  LHC.Prim.I# LHC.Prim.a#^84:LHC.Prim.I32 ← scrut^83:LHC.Prim.Int
  scrut^85 ← arg^82
  LHC.Prim.I# LHC.Prim.b#^86:LHC.Prim.I32 ← scrut^85:LHC.Prim.Int
  LHC.Prim.I# (LHC.Prim.*#
        LHC.Prim.a#^84 LHC.Prim.b#^86)
LHC.Prim.udiv# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.udiv# =
  λ arg^87:LHC.Prim.I32 arg^88:LHC.Prim.I32 → 
  primOut^89:LHC.Prim.I32 ← external udiv# arg^87 arg^88
  primOut^89
LHC.Prim.sdiv# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.sdiv# =
  λ arg^90:LHC.Prim.I32 arg^91:LHC.Prim.I32 → 
  primOut^92:LHC.Prim.I32 ← external sdiv# arg^90 arg^91
  primOut^92
LHC.Prim.urem# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.urem# =
  λ arg^93:LHC.Prim.I32 arg^94:LHC.Prim.I32 → 
  primOut^95:LHC.Prim.I32 ← external urem# arg^93 arg^94
  primOut^95
LHC.Prim.srem# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.srem# =
  λ arg^96:LHC.Prim.I32 arg^97:LHC.Prim.I32 → 
  primOut^98:LHC.Prim.I32 ← external srem# arg^96 arg^97
  primOut^98
LHC.Prim.sdiv : LHC.Prim.Int →  LHC.Prim.Int →  LHC.Prim.Int
LHC.Prim.sdiv =
  λ arg^99:LHC.Prim.Int arg^100:LHC.Prim.Int → 
  scrut^101 ← arg^99
  LHC.Prim.I# LHC.Prim.a#^102:LHC.Prim.I32 ← scrut^101:LHC.Prim.Int
  scrut^103 ← arg^100
  LHC.Prim.I# LHC.Prim.b#^104:LHC.Prim.I32 ← scrut^103:LHC.Prim.Int
  LHC.Prim.I# (LHC.Prim.sdiv#
        LHC.Prim.a#^102 LHC.Prim.b#^104)
LHC.Prim.srem : LHC.Prim.Int →  LHC.Prim.Int →  LHC.Prim.Int
LHC.Prim.srem =
  λ arg^105:LHC.Prim.Int arg^106:LHC.Prim.Int → 
  scrut^107 ← arg^105
  LHC.Prim.I# LHC.Prim.a#^108:LHC.Prim.I32 ← scrut^107:LHC.Prim.Int
  scrut^109 ← arg^106
  LHC.Prim.I# LHC.Prim.b#^110:LHC.Prim.I32 ← scrut^109:LHC.Prim.Int
  LHC.Prim.I# (LHC.Prim.srem#
        LHC.Prim.a#^108 LHC.Prim.b#^110)
LHC.Prim.mapM_ : ∀ sk_62_a. (sk_62_a →  LHC.Prim.IO ()) →  [sk_62_a] →  LHC.Prim.IO ()
LHC.Prim.mapM_ =
  λ @.sk_62_a^111:* LHC.Prim.fn^112:(sk_62_a →  LHC.Prim.IO ()) LHC.Prim.lst^113:[sk_62_a] → 
  case LHC.Prim.lst^113 of
    scrut^114:[sk_62_a]
    LHC.Prim.Nil  → 
      LHC.Prim.return
          @.LHC.Prim.Unit LHC.Prim.Unit
    LHC.Prim.Cons LHC.Prim.x^115:sk_62_a LHC.Prim.xs^116:[sk_62_a] → 
      LHC.Prim.thenIO
              @.LHC.Prim.Unit @.LHC.Prim.Unit
          (LHC.Prim.fn^112 LHC.Prim.x^115)
        (LHC.Prim.mapM_ @.sk_62_a^111
            LHC.Prim.fn^112 LHC.Prim.xs^116)
LHC.Prim.le# : LHC.Prim.I32 →  LHC.Prim.I32 →  LHC.Prim.I32
LHC.Prim.le# =
  λ arg^117:LHC.Prim.I32 arg^118:LHC.Prim.I32 → 
  primOut^119:LHC.Prim.I32 ← external le# arg^117 arg^118
  primOut^119
LHC.Prim.<= : LHC.Prim.Int →  LHC.Prim.Int →  LHC.Prim.Bool
LHC.Prim.<= =
  λ arg^120:LHC.Prim.Int arg^121:LHC.Prim.Int → 
  scrut^122 ← arg^120
  LHC.Prim.I# LHC.Prim.a#^123:LHC.Prim.I32 ← scrut^122:LHC.Prim.Int
  scrut^124 ← arg^121
  LHC.Prim.I# LHC.Prim.b#^125:LHC.Prim.I32 ← scrut^124:LHC.Prim.Int
  case LHC.Prim.le#
             LHC.Prim.a#^123
           LHC.Prim.b#^125 of
    scrut^126:LHC.Prim.I32
    '\NUL' → 
      LHC.Prim.False
    DEFAULT → 
      LHC.Prim.True
LHC.Prim.max : LHC.Prim.Int →  LHC.Prim.Int →  LHC.Prim.Int
LHC.Prim.max =
  λ LHC.Prim.a^127:LHC.Prim.Int LHC.Prim.b^128:LHC.Prim.Int → 
  case LHC.Prim.<= LHC.Prim.a^127
           LHC.Prim.b^128 of
    scrut^129:LHC.Prim.Bool
    LHC.Prim.True  → 
      LHC.Prim.b^128
    LHC.Prim.False  → 
      LHC.Prim.a^127
LHC.Prim.length : ∀ sk_76_a. [sk_76_a] →  LHC.Prim.Int
LHC.Prim.length =
  λ @.sk_76_a^130:* LHC.Prim.lst^131:[sk_76_a] → 
  LHC.Prim.length @.sk_76_a^130
    LHC.Prim.lst^131
LHC.Prim.emptyString : [LHC.Prim.Char]
LHC.Prim.emptyString =
  LHC.Prim.Nil
LHC.Prim.id : ∀ sk_79_a. sk_79_a →  sk_79_a
LHC.Prim.id =
  λ @.sk_79_a^132:* LHC.Prim.x^133:sk_79_a → 
  LHC.Prim.x^133
